arch/risc-v/src/bl602/bl602_netdev.c

```diff
@@ -0,0 +1,2113 @@
+   * packets.
+   */
+
+  bl602_net_addmac(dev, g_ipv6_ethallrouters.ether_addr_octet);
+
+#endif /* CONFIG_NET_ICMPv6_ROUTER */
+}
+#endif /* CONFIG_NET_ICMPv6 */
+
+static void scan_complete_indicate(void *data, void *param)
+{
+  int                        i;
+  struct scan_parse_param_s *para;
+  wifi_mgmr_scan_item_t *    scan;
+
+  para = (struct scan_parse_param_s *)data;
+  assert(para != NULL);
+  assert(para->priv != NULL);
+  para->priv->scan_result_len = 0;
+
+  for (i = 0;
+       i < sizeof(WIFI_MGMR.scan_items) / sizeof(WIFI_MGMR.scan_items[0]);
+       i++)
+    {
+      scan = &WIFI_MGMR.scan_items[i];
+
+      if (wifi_mgmr_scan_item_is_timeout(&WIFI_MGMR,
+                                         &(WIFI_MGMR.scan_items[i])))
+        {
+          scan->is_used = 0;
+        }
+      else if (scan->is_used)
+        {
+          if (para->flags & IW_SCAN_THIS_ESSID)
+            {
+              if (strncmp(scan->ssid,
+                          (char *)para->scan_req.essid,
+                          sizeof(scan->ssid)) == 0)
+                {
+                  scan->is_used = 1;
+                  para->priv->scan_result_len++;
+                }
+              else
+                {
+                  scan->is_used = 0;
+                }
+            }
+          else
+            {
+              para->priv->scan_result_len++;
+            }
+        }
+    }
+
+  sem_post(&g_wifi_scan_sem);
+  kmm_free(data);
+  return;
+}
+
+static int rssi_compare(const void *arg1, const void *arg2)
+{
+  wifi_mgmr_scan_item_t *item1 = &WIFI_MGMR.scan_items[*(uint8_t *)arg1];
+  wifi_mgmr_scan_item_t *item2 = &WIFI_MGMR.scan_items[*(uint8_t *)arg2];
+
+  return item1->rssi - item2->rssi;
+}
+
+static int format_scan_result_to_wapi(struct iwreq *req, int result_cnt)
+{
+  int      i = 0;
+  int      j = 0;
+  int      event_buff_len = 0;
+  uint8_t *curr_pos       = NULL;
+
+  uint8_t *rssi_list = NULL; /* for sort */
+
+  if (result_cnt == 0)
+    {
+      return OK;
+    }
+
+  /* More compact arrangement */
+
+  event_buff_len = result_cnt *
+    (offsetof(struct iw_event, u) * 4 + sizeof(struct sockaddr) +
+    sizeof(struct iw_freq) + sizeof(struct iw_quality) +
+    sizeof(struct iw_point) + IW_ESSID_MAX_SIZE);
+
+  if (req->u.data.length == 0 || req->u.data.length < event_buff_len)
+    {
+      return -E2BIG;
+    }
+
+  req->u.data.length = event_buff_len;
+
+  /* alloc rssi list */
+
+  rssi_list = kmm_malloc(result_cnt * sizeof(uint8_t));
+  if (rssi_list == NULL)
+    {
+      return -ENOMEM;
+    }
+
+  /* record all valid scan result items */
+
+  for (i = 0, j = 0;
+       i < sizeof(WIFI_MGMR.scan_items) / sizeof(WIFI_MGMR.scan_items[0]);
+       i++)
+    {
+      wifi_mgmr_scan_item_t *scan = &WIFI_MGMR.scan_items[i];
+
+      if (scan->is_used == 0)
+        {
+          continue;
+        }
+
+      rssi_list[j++] = i;
+    }
+
+  assert(j == result_cnt);
```

Comment:
How would this ever fail?

Reply:
I am worried that in the firmware, there will be new modifications to the results between the completion of the scan and the time I get the scan results. At least I have not encountered its assertion failure. This is added when debugging the code. I will change it to DEBUG_ASSERT.